src/drivers/net/mcst/l_e1000.c

```diff
@@ -0,0 +1,728 @@
+#define E1000_LOG_TX_BUFFERS       OPTION_GET(NUMBER,log_tx_desc_quantity)
+
+/*
+ * Set the number of Tx and Rx buffers, using Log_2(# buffers).
+ * Reasonable default values are 16 Tx buffers, and 256 Rx buffers.
+ * That translates to 4 (16 == 2^^4) and 8 (256 == 2^^8).
+ */
+
+#define TX_RING_SIZE            (1 << (E1000_LOG_TX_BUFFERS))
+#define TX_RING_MOD_MASK        (TX_RING_SIZE - 1)
+#define TX_RING_LEN_BITS        ((E1000_LOG_TX_BUFFERS) << 12)
+
+#define RX_RING_SIZE            (1 << (E1000_LOG_RX_BUFFERS))
+#define RX_RING_MOD_MASK        (RX_RING_SIZE - 1)
+#define RX_RING_LEN_BITS        ((E1000_LOG_RX_BUFFERS) << 4)
+
+struct l_e1000_rx_desc {
+	uint32_t    base;
+	int16_t     buf_length;
+	int16_t     status;
+	int16_t     msg_length;
+	uint16_t    reserved1;
+	uint32_t    etmr;
+} __attribute__((packed));
+
+struct l_e1000_tx_desc {
+	uint32_t    base;
+	int16_t     buf_length;
+	int16_t     status;
+	uint32_t    misc;
+	uint32_t    etmr;
+} __attribute__((packed));
+
+#if 0
+struct rx_buf {
+	uint8_t b[0x1000] __attribute__((aligned(8)));
+};
+
+#endif
+
+struct l_e1000_dma_area {
+	struct l_e1000_init_block init_block __attribute__((aligned(32)));
+	struct l_e1000_rx_desc    rx_ring[RX_RING_SIZE] __attribute__((aligned(16)));
+	struct l_e1000_tx_desc    tx_ring[TX_RING_SIZE] __attribute__((aligned(16)));
+	//struct rx_buf      rx_buf[RX_RING_SIZE]    __attribute__((aligned(8)));
+};
+
+struct mii_if {
+	unsigned char full_duplex;
+	unsigned char supports_gmii;
+	unsigned char phy_id_mask;
+	unsigned char reg_num_mask;
+	unsigned char phy_id;
+};
+
+static struct l_e1000_dma_area e1000_dma_area __attribute__((aligned(32)));
+
+struct l_e1000_priv {
+	struct net_device *netdev;
+	struct rx_buf             *rx_buf;
+
+	struct sk_buff *rx_skbs[RX_RING_SIZE];
+
+	struct l_e1000_init_block *init_block;
+	struct l_e1000_rx_desc    *rx_ring;
+	struct l_e1000_tx_desc    *tx_ring;
+
+	void                      *dma_area;
+	uintptr_t                  dma_addr;
+
+	uintptr_t base_ioaddr; /* iomapped device regs */
+
+	struct mii_if mii_if;
+	int cur_rx;
+};
+
+
+static inline
+uint32_t e1000_read_e_csr(struct l_e1000_priv *ep) {
+	uint32_t val = 0;
+	if (ep->base_ioaddr) {
+		val = e2k_read32(ep->base_ioaddr + E_CSR);
+		return val;
+	}
+	return 0xabcdefab;
+}
+
+static inline
+void e1000_write_e_csr(struct l_e1000_priv *ep, int val) {
+	if (ep->base_ioaddr) {
+		e2k_write32(val, ep->base_ioaddr + E_CSR);
+	}
+}
+
+static inline
+uint32_t e1000_read_mgio_csr(struct l_e1000_priv *ep) {
+	if (ep->base_ioaddr) {
+		return e2k_read32(ep->base_ioaddr + MGIO_CSR);
+	}
+	return 0xabcdefab;
+}
+
+static inline
+void e1000_write_mgio_csr(struct l_e1000_priv *ep, int val) {
+	if (ep->base_ioaddr) {
+		e2k_write32(val, ep->base_ioaddr + MGIO_CSR);
+	}
+}
+
+static inline
+uint32_t e1000_read_mgio_data(struct l_e1000_priv *ep) {
+	if (ep->base_ioaddr) {
+		return e2k_read32(ep->base_ioaddr + MGIO_DATA);
+	}
+	return 0xabcdefab;
+}
```

Comment:
Why `0xabcdefab`? It looks weird.

Reply:
The same is for e1000_read_e_csr, e1000_read_mgio_csr, ... etc.

Reply:
done